README.md
===============================================================================
Overview
--------
MyBatis Generator (MBG) is a code generator for the MyBatis (and iBATIS) SQL
mapping framefork.  MBG will introspect database tables (through JDBC
DatabaseMetaData) and generate SQL Map XML files, Java model object (POJOs)
that match the table, and (optionally) Java client classes that use the other
generated objects.

For full documentation, please refer to the user's manual at
doc/html/index.html in this distribution.

Dependencies
------------
There are no dependencies beyond the JRE.  JRE 5.0 or above is required.
Also required is a JDBC driver that implements the DatabaseMetaData interface,
especially the "getColumns" and "getPrimaryKeys" methods.

Support
-------
Support is provided through the user mailing list.  Mail
questions or bug reports to:

  mybatis-user@googlegroups.com


doc/ReleaseNotes.txt
-------------------------------------------------------------------------------
Version 1.3.1

Release notes are no longer maintained in this file.  See the "What's New"
page in MyBatis Generator documentation for change details.

-------------------------------------------------------------------------------
Version 1.3.0

Announcements:

1. Moved to mybatis.org, renamed MyBatis Generator


-------------------------------------------------------------------------------
Version 1.2.2 (Never Released):

Announcements:
1. CommentGenerator interface has changed
2. Ibator is now built with Maven
3. Added support for iBATIS version 3

Bugs:
1. Fixed NPE when no DAOs are generated
2. Fixed IBATIS-579 - don't allow column names with spaces to break
   across lines
3. Fixed NPE and incorrect calculation in generated equals when certain
   fields are null - thanks to Benjamin Klatt for finding this bug
4. Fixed IBATIS-601 - improper validation of <generatedKey>
5. Fixed IBATIS-609 - incorrect parsing of Java generic types
6. Fixed spelling error LONCVARCHAR->LONGVARCHAR (thanks Allard)
7. Fixed IBATIS-731 - change name of primary key variable to avoid conflicts
8. Fixed IBATIS-699 - Overwrite unmergeable XML files if enabled
9. Fixed issue where insertSelective failed if there is a sequence generating the
   primary key (issue only with iBATIS3)

Enhancements:
1. IBATIS-569 - Make it easier to override IbatorRules in plugins
2. IBATIS-571 - Added "autoDelimitKeywords" support to <ibatorContext>
3. IBATIS-577 - Define SQL fragments for column lists.  Thanks to Iwao
   AVE! for the idea and the initial patch
4. Added some additional command line arguments
5. Added logging
6. Added Case Insensitive LIKE example Plugin
7. Added delimitAllCoumns to table configurations
8. Added page describing some of PostgreSQL idiosyncracies with Ibator
9. IBATIS-586 - Added the ability to specify nested property elements on
   columnOverrides (thanks to Dan Turkenkopf for the idea and a nice initial patch)
10. Gather column comments during introspection
11. IBATIS-592 - Additional attributes for SqlMaps
12. Fixed addCriterionforJDBC* methods so that they all do a null check
13. Fixed IbatorRunner so that configuration errors are shown (thanks to Karel Rank)
14. Added XML ids to changeable attributes on introspected table, also
    change the XML merger so that the XML ids no longer need to contain
    the prefix "ibatorgenerated_"
15. Many refactorings to make way for iBATIS3
16. IBATIS-605 - Added Informix Dialect
17. Added ability to set annotations on method parameters
18. Addedd support for "distinct" on select by example methods
19. Added new "or" method to example classes
20. Added new "useCompoundPropertyNames" property on <table>
21. Further improved extensibility of the example classes
22. EqulasHashCodePlugin now generates far superior methods


-------------------------------------------------------------------------------
Version 1.2.1:

Bugs:
1. Fixed IbatorObjectFactory so that it will find internal classes on the
   context classloader.
2. Fixed IBATIS-565 - ill formed comment in the SqlMapConfigPlugin.

Enhancements:
1. Modified plugin methods for model fields, getters, and setters so that
   the plugin will know which type of class (Primary Key, Base Record, or
   Record with BLOBs) is being generated. 
2. Added methods to IntrospectedTable to get/set attributes.  This 
   allows plugin classes to maintain table based state between
   plugin calls.
3. Added "initialized" method to the plugin API.  This
   allows plugins to alter some of the fundamental code generation items
   (like the name of a generated class, for example).
4. Added an example plugin to show usage of the "initialized" method.

-------------------------------------------------------------------------------
Version 1.2.0:  Renamed to Ibator

Bugs:
1. Fixed the JavaTypeResolver so that an unsupported data type can be
   overridden by configuration.
2. Fixed IBATIS-523 - bug in EqualsHashCodePlugin
3. Fixed IBATIS-542 - upgrade build to Ant 1.7.1

Enhancements:
1. Ibator now includes a plugin machanism that you can use to extend or
   modify code generated by Ibator.  In most cases, you should be able
   to write plugins to replace any custom classes that subclass one of
   ibator's built-in generators.  Plugins are now the preferred method for
   extending Ibator and we recommend that you do not extend Ibator's
   built in generators unless absolutely necessary.
2. Major refactoring of Ibator internals.  This has changed
   the interface for overriding any of the supplied Ibator generators.
3. Added support for runtimeCatalog and runtimeSchema properties to the
   table configuration.  Thanks to Dan Turkenkopf for the idea and the
   patch!
4. New method - insertSelective
   
   
Changes:
1. JSE 5.0 or greater is required
2. Changed and centralized classloading - moved the <classPathEntry> from
   the <jdbcConnection> element to <ibatorConfiguration>, removed the
   rootClasspath property from <javaModelGenerator>
3. Changed API on the JavaTypeResolver interface
4. Removed Legacy generator set
5. Removed after generation hook methods in favor of the new ibator
   plugin mechanism.
6. <ibatorContext> elements now require an id

-------------------------------------------------------------------------------
Version 1.1.0:
Bugs - 
1. corner case where the new "by example" methods fail if a criteria
   class has been added to the list, but there are no criteria set
2. trimStrings on JavaModelGenerator not working
3. Fix the XML file merger so that entities are preserved
4. Fixed the configuration parser so that it correctly deals with
   external entities
5. IBATIS-434 - Incorrect default mapping for BIT datatype
6. Fixed bug where Abator generated incorrect properties for certain
   database columns (for example, if the column name is I_NAME)

Enhancements - 

src/main/java/org/mybatis/generator/internal/types/JavaTypeResolverDefaultImpl.java
/*
 *  Copyright 2005 The Apache Software Foundation
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.mybatis.generator.internal.types;

import java.math.BigDecimal;
import java.sql.Types;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Properties;

import org.mybatis.generator.api.IntrospectedColumn;
import org.mybatis.generator.api.JavaTypeResolver;
import org.mybatis.generator.api.dom.java.FullyQualifiedJavaType;
import org.mybatis.generator.config.Context;
import org.mybatis.generator.config.PropertyRegistry;
import org.mybatis.generator.internal.util.StringUtility;

/**
 * 
 * @author Jeff Butler
 */
public class JavaTypeResolverDefaultImpl implements JavaTypeResolver {

	protected List<String> warnings;

	protected Properties properties;

	protected Context context;

	protected boolean forceBigDecimals;

	protected Map<Integer, JdbcTypeInformation> typeMap;

	// these are added manually until we move to JDK 6
	// TODO - remove after JDK 6 and use the java.sql.Types constants instead
	static final int NVARCHAR = -9;
	static final int NCHAR = -15;
	static final int NCLOB = 2011;

	public JavaTypeResolverDefaultImpl() {
		super();
		properties = new Properties();
		typeMap = new HashMap<Integer, JdbcTypeInformation>();

		typeMap.put(Types.ARRAY, new JdbcTypeInformation("ARRAY", //$NON-NLS-1$
				new FullyQualifiedJavaType(Object.class.getName())));
		typeMap.put(Types.BIGINT, new JdbcTypeInformation("BIGINT", //$NON-NLS-1$
				new FullyQualifiedJavaType(Long.class.getName())));
		typeMap.put(Types.BINARY, new JdbcTypeInformation("BINARY", //$NON-NLS-1$
				new FullyQualifiedJavaType("byte[]"))); //$NON-NLS-1$
		typeMap.put(Types.BIT, new JdbcTypeInformation("BIT", //$NON-NLS-1$
				new FullyQualifiedJavaType(Boolean.class.getName())));
		typeMap.put(Types.BLOB, new JdbcTypeInformation("BLOB", //$NON-NLS-1$
				new FullyQualifiedJavaType("byte[]"))); //$NON-NLS-1$
		typeMap.put(Types.BOOLEAN, new JdbcTypeInformation("BOOLEAN", //$NON-NLS-1$
				new FullyQualifiedJavaType(Boolean.class.getName())));
		typeMap.put(Types.CHAR, new JdbcTypeInformation("CHAR", //$NON-NLS-1$
				new FullyQualifiedJavaType(String.class.getName())));
		typeMap.put(Types.CLOB, new JdbcTypeInformation("CLOB", //$NON-NLS-1$
				new FullyQualifiedJavaType(String.class.getName())));
		typeMap.put(Types.DATALINK, new JdbcTypeInformation("DATALINK", //$NON-NLS-1$
				new FullyQualifiedJavaType(Object.class.getName())));
		typeMap.put(Types.DATE, new JdbcTypeInformation("DATE", //$NON-NLS-1$
				new FullyQualifiedJavaType(Date.class.getName())));
		typeMap.put(Types.DISTINCT, new JdbcTypeInformation("DISTINCT", //$NON-NLS-1$
				new FullyQualifiedJavaType(Object.class.getName())));
		typeMap.put(Types.DOUBLE, new JdbcTypeInformation("DOUBLE", //$NON-NLS-1$
				new FullyQualifiedJavaType(Double.class.getName())));
		typeMap.put(Types.FLOAT, new JdbcTypeInformation("FLOAT", //$NON-NLS-1$
				new FullyQualifiedJavaType(Double.class.getName())));
		typeMap.put(Types.INTEGER, new JdbcTypeInformation("INTEGER", //$NON-NLS-1$
				new FullyQualifiedJavaType(Integer.class.getName())));
		typeMap.put(Types.JAVA_OBJECT, new JdbcTypeInformation("JAVA_OBJECT", //$NON-NLS-1$
				new FullyQualifiedJavaType(Object.class.getName())));
		typeMap.put(Types.LONGVARBINARY, new JdbcTypeInformation("LONGVARBINARY", //$NON-NLS-1$
				new FullyQualifiedJavaType("byte[]"))); //$NON-NLS-1$
		typeMap.put(Types.LONGVARCHAR, new JdbcTypeInformation("LONGVARCHAR", //$NON-NLS-1$
				new FullyQualifiedJavaType(String.class.getName())));
		typeMap.put(NCHAR, new JdbcTypeInformation("NCHAR", //$NON-NLS-1$
				new FullyQualifiedJavaType(String.class.getName())));
		typeMap.put(NCLOB, new JdbcTypeInformation("NCLOB", //$NON-NLS-1$
				new FullyQualifiedJavaType(String.class.getName())));
		typeMap.put(NVARCHAR, new JdbcTypeInformation("NVARCHAR", //$NON-NLS-1$
				new FullyQualifiedJavaType(String.class.getName())));
		typeMap.put(Types.NULL, new JdbcTypeInformation("NULL", //$NON-NLS-1$
				new FullyQualifiedJavaType(Object.class.getName())));
		typeMap.put(Types.OTHER, new JdbcTypeInformation("OTHER", //$NON-NLS-1$
				new FullyQualifiedJavaType(Object.class.getName())));
		typeMap.put(Types.REAL, new JdbcTypeInformation("REAL", //$NON-NLS-1$
				new FullyQualifiedJavaType(Float.class.getName())));
		typeMap.put(Types.REF, new JdbcTypeInformation("REF", //$NON-NLS-1$
				new FullyQualifiedJavaType(Object.class.getName())));
		typeMap.put(Types.SMALLINT, new JdbcTypeInformation("SMALLINT", //$NON-NLS-1$
				new FullyQualifiedJavaType(Short.class.getName())));
		typeMap.put(Types.STRUCT, new JdbcTypeInformation("STRUCT", //$NON-NLS-1$
				new FullyQualifiedJavaType(Object.class.getName())));
		typeMap.put(Types.TIME, new JdbcTypeInformation("TIME", //$NON-NLS-1$
				new FullyQualifiedJavaType(Date.class.getName())));
		typeMap.put(Types.TIMESTAMP, new JdbcTypeInformation("TIMESTAMP", //$NON-NLS-1$
				new FullyQualifiedJavaType(Date.class.getName())));
		typeMap.put(Types.TINYINT, new JdbcTypeInformation("TINYINT", //$NON-NLS-1$
				new FullyQualifiedJavaType(Byte.class.getName())));
		typeMap.put(Types.VARBINARY, new JdbcTypeInformation("VARBINARY", //$NON-NLS-1$
				new FullyQualifiedJavaType("byte[]"))); //$NON-NLS-1$
		typeMap.put(Types.VARCHAR, new JdbcTypeInformation("VARCHAR", //$NON-NLS-1$
				new FullyQualifiedJavaType(String.class.getName())));

	}

	public void addConfigurationProperties(Properties properties) {
		this.properties.putAll(properties);
		forceBigDecimals = StringUtility.isTrue(properties
				.getProperty(PropertyRegistry.TYPE_RESOLVER_FORCE_BIG_DECIMALS));
	}

	public FullyQualifiedJavaType calculateJavaType(IntrospectedColumn introspectedColumn) {
		FullyQualifiedJavaType answer;
		JdbcTypeInformation jdbcTypeInformation = typeMap.get(introspectedColumn.getJdbcType());

		if (jdbcTypeInformation == null) {
			switch (introspectedColumn.getJdbcType()) {
			case Types.DECIMAL:
			case Types.NUMERIC:
				if (introspectedColumn.getScale() > 0 || introspectedColumn.getLength() > 18 || forceBigDecimals) {
					answer = new FullyQualifiedJavaType(BigDecimal.class.getName());
				} else if (introspectedColumn.getLength() > 9) {
					answer = new FullyQualifiedJavaType(Long.class.getName());
				} else if (introspectedColumn.getLength() > 4) {
					answer = new FullyQualifiedJavaType(Integer.class.getName());
				} else {
					answer = new FullyQualifiedJavaType(Short.class.getName());
				}
				break;

			default:
				answer = null;
				break;
			}
		} else {
			answer = jdbcTypeInformation.getFullyQualifiedJavaType();
		}

		return answer;
	}

	public String calculateJdbcTypeName(IntrospectedColumn introspectedColumn) {
		String answer;
		JdbcTypeInformation jdbcTypeInformation = typeMap.get(introspectedColumn.getJdbcType());

		if (jdbcTypeInformation == null) {
			switch (introspectedColumn.getJdbcType()) {
			case Types.DECIMAL:
				answer = "DECIMAL"; //$NON-NLS-1$
				break;
			case Types.NUMERIC:
				answer = "NUMERIC"; //$NON-NLS-1$
				break;
			default:
				answer = null;
				break;
			}
		} else {
			answer = jdbcTypeInformation.getJdbcTypeName();
		}

		return answer;
	}

	public void setWarnings(List<String> warnings) {
		this.warnings = warnings;
	}

	public void setContext(Context context) {
		this.context = context;
	}

	private static class JdbcTypeInformation {
		private String jdbcTypeName;

		private FullyQualifiedJavaType fullyQualifiedJavaType;

		public JdbcTypeInformation(String jdbcTypeName, FullyQualifiedJavaType fullyQualifiedJavaType) {
			this.jdbcTypeName = jdbcTypeName;
			this.fullyQualifiedJavaType = fullyQualifiedJavaType;
		}

		public String getJdbcTypeName() {
			return jdbcTypeName;
		}

		public FullyQualifiedJavaType getFullyQualifiedJavaType() {
			return fullyQualifiedJavaType;
		}
	}
}
